### PR TITLE
Conf tree

### PR DESCRIPTION
**Pull request by oliver-sanders**
**Sibling PRs:**

* https://github.com/cylc/cylc-flow/pull/3559
* https://github.com/cylc/cylc-sphinx-extensions/pull/22

**Highlights:**

* Use the cylc-sphinx-extensions library (remove old built-in extensions).
* Autodocument the Cylc suite and global configurations.
* Fix broken references.

**Note:**

The presentation is down to the Sphinx theme not the documentation method.

**Further Work:**

* For configs shared between the suite and global configs, I've only documented one copy. Once the platforms work is done I'll work out a way of automatically referencing one from the other.
* The docs still require reviewing, I've just copied them from a to b and fixed things up a little.
* I've only corrected the broken references for now, ultimately we will need to go through the code making changes like this:

    ```diff
     - See the ``[scheduling]`` section.
     + See the :cylc:conf:`[scheduling]` section.
    ```

**Benefits:**

* Consistency with Cylc Flow, obvious.
* Integrity checking, the docs will fail to build if a configuration is removed, but still referenced in the docs.
* Links, all references to cylc sections result in hyperlinks to the relevant content.
* Collaboration, objects can be shared with other projects (e.g. we can reference Cylc configurations from the Rose docs).

**Screenshots**

Screenshot 2020-04-29 at 11 51 42
Screenshot 2020-04-29 at 11 52 15
Screenshot 2020-04-29 at 11 52 46

### CODE REVIEW

**Comment by oliver-sanders** [comment]
(test failure expected).

File: setup.py [hunk 1]
```diff
@@ -39,11 +39,12 @@ def get_version(module, path):
 )
 
 INSTALL_REQUIRES = [
-    'sphinx==2.0.*',
+    'sphinx>=2.0.*,<3.0.0',
```
Comment (datamel) [inline setup.py:42]: Just wondering.... is there a reason why it wouldn't work with a strict version of sphinx>2.0.*, is there something in sphinx>=3.0.0 that would stop it working?

Comment (oliver-sanders) [reply setup.py:42]: Sooooo,

These docs do not have any especially strict coupling to Sphinx, we aren't using bleeding edge features or anything like that.

However, the dependent cylc-sphinx-extensions, by it's nature as a Sphinx extension has much tighter coupling. So by making the sphinx requirement vague in cylc-doc we are effectively allowing cylc-sphinx-extensions to pick the version, which, unfortunately, is necessary due to a downstream dependency being broken at Sphinx 2.4 😞.

**Comment by oliver-sanders** [comment]
Rebased, the tests should fail (because the sibling PRs aren't merged) but they should have the potential to pass.

File: src/user-guide/writing-suites.rst [hunk 2]
```diff
@@ -3624,18 +3629,19 @@ commenting them out of the graph.
...
 validity of the tasks and their dependencies, but they are not actually
 loaded into the suite at start-up. Other tasks that depend on the
 omitted ones, if any, will have to wait on their insertion at a later
 time or otherwise be triggered manually.
 
 Finally, with Jinja2 (:ref:`Jinja`) you can radically alter
-suite structure by including or excluding tasks from the ``[scheduling]``
-and ``[runtime]`` sections according to the value of a single logical flag
-defined at the top of the suite.
+suite structure by including or excluding tasks from th
```
Comment (datamel) [inline src/user-guide/writing-suites.rst:3642]:
```suggestion
suite structure by including or excluding tasks from the
```

**Comment by oliver-sanders** [comment]
I've not seen that before, usually Sphinx dumps traceback into a file, sadly not in your case.

Could you fetch some diagnostics, namely `pip list`.

There are [some dependencies](https://github.com/cylc/cylc-doc/blob/master/.github/workflows/test.yml#L24-L26) pip can't install, I wouldn't imagine their absence could cause a segfault but worth noting.

I would build an env by doing:

```
conda install pygraphviz
pip install -e ~/cylc-flow/
pip install -e ~/cylc-sphinx-extensions/
pip install -e ~/cylc-doc/
```

**Comment by kinow** [comment]
>Could you fetch some diagnostics, namely pip list.

<details>
  <summary>pip list</summary>

```
(venv) kinow@ranma:~/Development/python/workspace/cylc-doc$ pip list
Package                        Version   Location
------------------------------ --------- ---------------------------------------------------------------
alabaster                      0.7.12
aniso8601                      6.0.0
ansimarkup                     1.4.0
attrs                          19.3.0
Babel                          2.7.0
certifi                        2019.6.16
chardet                        3.0.4
Click                          7.0
codecov                        2.0.22
colorama                       0.4.1
coverage                       5.0.4
cycler                         0.10.0
cylc-doc                       0.0.1     /home/kinow/Development/python/workspace/cylc-doc
cylc-flow                      8.0a1     /home/kinow/Development/python/workspace/cylc-flow
cylc-sphinx-extensions         1.0.0     /home/kinow/Development/python/workspace/cylc-sphinx-extensions
docutils                       0.15.1
ecdsa                          0.13.2
empy                           3.3.4
ERAlchemy                      1.2.10
future                         0.17.1
graphene                       2.1.6
graphql-core                   2.2.1
graphql-relay                  0.4.5
hieroglyph                     1.0.0
idna                           2.8
imagesize                      1.1.0
importlib-metadata             1.6.0
Jinja2                         2.11.1
kiwisolver                     1.2.0
MarkupSafe                     1.1.1
matplotlib                     3.2.1
metomi-isodatetime             1!2.0.1
more-itertools                 8.2.0
numpy                          1.18.2
packaging                      19.0
pandas                         0.25.3
Pillow                         7.1.1
pip                            20.1
pluggy                         0.13.1
promise                        2.2.1
protobuf                       3.11.3
psutil                         5.7.0
py                             1.8.1
pyasn1                         0.4.5
pycodestyle                    2.5.0
Pygments                       2.4.2
pygraphviz                     1.5
Pympler                        0.8
pyparsing                      2.4.1.1
pytest                         5.3.5
pytest-cov                     2.8.1
python-dateutil                2.8.1
python-jose                    3.0.1
pytz                           2019.1
pyzmq                          18.1.1
requests                       2.22.0
rsa                            4.0
Rx                             1.6.1
setuptools                     41.0.1
six                            1.12.0
snowballstemmer                1.9.0
Sphinx                         2.3.1
sphinx-rtd-theme               0.4.3
sphinxcontrib-applehelp        1.0.1
sphinxcontrib-devhelp          1.0.1
sphinxcontrib-htmlhelp         1.0.2
sphinxcontrib-jsmath           1.0.1
sphinxcontrib-qthelp           1.0.2
sphinxcontrib-serializinghtml  1.1.3
sphinxcontrib-svg2pdfconverter 0.1.0
SQLAlchemy                     1.3.8
testfixtures                   6.11.0
urllib3                        1.25.3
urwid                          2.1.0
wcwidth                        0.1.9
wheel                          0.33.4
zipp                           3.1.0
```
</details>

**Comment by kinow** [comment]
>There are some dependencies pip can't install, I wouldn't imagine their absence could cause a segfault but worth noting.

These dependencies seem fine, as I use graphviz in other projects and haven't had issues with pip so far. 

```
(venv) kinow@ranma:~/Development/python/workspace/cylc-doc$ sudo apt-get install -y graphviz pkg-config libgraphviz-dev
[sudo] password for kinow: 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
pkg-config is already the newest version (0.29.1-0ubuntu2).
graphviz is already the newest version (2.40.1-2).
libgraphviz-dev is already the newest version (2.40.1-2).
0 upgraded, 0 newly installed, 0 to remove and 41 not upgraded.
(venv) kinow@ranma:~/Development/python/workspace/cylc-doc$ pip install pygraphviz
Requirement already satisfied: pygraphviz in ./venv/lib/python3.7/site-packages (1.5)
```

**Comment by kinow** [comment]
The `make html` appears to result in `sphinx-build -M html src doc/8.0a1 -n`. So `gdb python`, followed by running this command:

<details>
<summary>gdb output</summary>

```
(venv) kinow@ranma:~/Development/python/workspace/cylc-doc$ gdb python
GNU gdb (Ubuntu 8.1-0ubuntu3.2) 8.1.0.20180409-git
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from python...done.
(gdb) r /home/kinow/Development/python/workspace/cylc-doc/venv/bin/sphinx-build -M html src doc/8.0a1 -n
Starting program: /home/kinow/Development/python/workspace/cylc-doc/venv/bin/python /home/kinow/Development/python/workspace/cylc-doc/venv/bin/sphinx-build -M html src doc/8.0a1 -n
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Running Sphinx v2.3.1
[autosummary] generating autosummary for: appendices/command-ref.rst, glossary.rst, index.rst, installation.rst, introduction.rst, reference/api/index.rst, reference/api/zmq.rst, reference/config/global.rst, reference/config/index.rst, reference/config/suite.rst, ..., user-guide/plugins/main-loop/built-in/cylc.flow.main_loop.log_memory.rst, user-guide/plugins/main-loop/index.rst, user-guide/remote-job-management.rst, user-guide/running-suites.rst, user-guide/suite-name-reg.rst, user-guide/suite-storage-etc.rst, user-guide/task-implementation.rst, user-guide/task-job-submission.rst, user-guide/workflows.rst, user-guide/writing-suites.rst
[New Thread 0x7fffee1c5700 (LWP 5719)]
[New Thread 0x7fffed9c4700 (LWP 5720)]
[New Thread 0x7fffeb1c3700 (LWP 5721)]
[New Thread 0x7fffe69c2700 (LWP 5722)]
[New Thread 0x7fffe41c1700 (LWP 5723)]
[New Thread 0x7fffe19c0700 (LWP 5724)]
[New Thread 0x7fffdf1bf700 (LWP 5725)]
[New Thread 0x7fffd919a700 (LWP 5726)]
loading intersphinx inventory from http://metomi.github.io/rose/doc/html/objects.inv...
[New Thread 0x7fffd8999700 (LWP 5727)]
loading intersphinx inventory from https://docs.python.org/objects.inv...
intersphinx inventory has moved: https://docs.python.org/objects.inv -> https://docs.python.org/3/objects.inv
[Thread 0x7fffd8999700 (LWP 5727) exited]
[Thread 0x7fffd919a700 (LWP 5726) exited]

Thread 1 "python" received signal SIGSEGV, Segmentation fault.
0x00007ffff581d962 in simplequeue_dealloc () at /usr/local/src/conda/python-3.7.6/Modules/_queuemodule.c:29
29	/usr/local/src/conda/python-3.7.6/Modules/_queuemodule.c: No such file or directory.
(gdb) 
(gdb) 
(gdb) 
(gdb) bt
#0  0x00007ffff581d962 in simplequeue_dealloc () at /usr/local/src/conda/python-3.7.6/Modules/_queuemodule.c:29
#1  0x00005555556ba11b in delitem_common (old_value=0x7ffff29d7ea8, ix=<optimized out>, hash=<optimized out>, mp=<optimized out>)
    at /tmp/build/80754af9/python_1553721932202/work/Objects/dictobject.c:1504
#2  _PyDict_DelItem_KnownHash () at /tmp/build/80754af9/python_1553721932202/work/Objects/dictobject.c:1556
#3  0x0000555555726951 in _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:1591
#4  0x000055555566a4f9 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3930
#5  0x000055555566b5d5 in _PyFunction_FastCallDict () at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:376
#6  0x0000555555682f3e in object_vacall () at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:1198
#7  0x0000555555683089 in PyObject_CallFunctionObjArgs (callable=<optimized out>) at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:1263
#8  0x00005555556e11df in handle_callback (callback=0x7ffff29b6f28, ref=<optimized out>) at /tmp/build/80754af9/python_1553721932202/work/Objects/weakrefobject.c:872
#9  PyObject_ClearWeakRefs () at /tmp/build/80754af9/python_1553721932202/work/Objects/weakrefobject.c:953
#10 0x00005555557018d7 in subtype_dealloc () at /tmp/build/80754af9/python_1553721932202/work/Objects/typeobject.c:1201
#11 0x00005555556e1977 in set_dealloc () at /tmp/build/80754af9/python_1553721932202/work/Objects/setobject.c:568
#12 0x0000555555663a27 in dict_dealloc () at /tmp/build/80754af9/python_1553721932202/work/Objects/dictobject.c:1905
#13 0x0000555555701971 in subtype_dealloc () at /tmp/build/80754af9/python_1553721932202/work/Objects/typeobject.c:1241
#14 0x0000555555652c08 in frame_dealloc () at /tmp/build/80754af9/python_1553721932202/work/Objects/frameobject.c:470
#15 0x000055555566b542 in function_code_fastcall (globals=<optimized out>, nargs=<optimized out>, args=<optimized out>, co=0x7ffff34edae0)
    at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:291
#16 _PyFunction_FastCallDict () at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:322
#17 0x0000555555726250 in do_call_core (kwdict=0x0, callargs=0x7ffff2a41128, func=0x7ffff348bb70) at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:4645
#18 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3191
#19 0x000055555566a4f9 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3930
#20 0x00005555556cca27 in _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:433
#21 0x0000555555724ad0 in call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:4616
#22 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3110
#23 0x00005555556cc79b in function_code_fastcall (globals=<optimized out>, nargs=1, args=<optimized out>, co=<optimized out>)
    at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:283
#24 _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:408
#25 0x0000555555724ad0 in call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:4616
#26 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3110
#27 0x000055555566a4f9 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3930
#28 0x000055555566b5d5 in _PyFunction_FastCallDict () at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:376
#29 0x0000555555682cbe in _PyObject_Call_Prepend () at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:904
#30 0x00005555556c50aa in slot_tp_init () at /tmp/build/80754af9/python_1553721932202/work/Objects/typeobject.c:6613
#31 0x00005555556cd9b8 in type_call (kwds=0x0, args=0x7ffff5226cc8, type=<optimized out>) at /tmp/build/80754af9/python_1553721932202/work/Objects/typeobject.c:949
#32 _PyObject_FastCallKeywords () at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:199
#33 0x0000555555729026 in call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:4619
#34 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3124
#35 0x000055555566a4f9 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3930
#36 0x00005555556cc9c5 in _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:433
#37 0x0000555555724846 in call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>)
---Type <return> to continue, or q <return> to quit---
    at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:4616
#38 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3124
#39 0x000055555566a4f9 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3930
#40 0x00005555556cc9c5 in _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:433
#41 0x0000555555724ad0 in call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:4616
#42 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3110
#43 0x00005555556cc79b in function_code_fastcall (globals=<optimized out>, nargs=1, args=<optimized out>, co=<optimized out>)
    at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:283
#44 _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:408
#45 0x0000555555728ed9 in call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:4616
#46 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3093
#47 0x000055555566a4f9 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3930
#48 0x00005555556cc9c5 in _PyFunction_FastCallKeywords () at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:433
#49 0x0000555555724846 in call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:4616
#50 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3124
#51 0x00005555556cc620 in function_code_fastcall () at /tmp/build/80754af9/python_1553721932202/work/Objects/call.c:283
#52 0x0000555555724846 in call_function (kwnames=0x0, oparg=<optimized out>, pp_stack=<synthetic pointer>)
    at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:4616
#53 _PyEval_EvalFrameDefault () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3124
#54 0x000055555566a4f9 in _PyEval_EvalCodeWithName () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3930
#55 0x000055555566b3c4 in PyEval_EvalCodeEx () at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:3959
#56 0x000055555566b3ec in PyEval_EvalCode (co=<optimized out>, globals=<optimized out>, locals=<optimized out>)
    at /tmp/build/80754af9/python_1553721932202/work/Python/ceval.c:524
#57 0x0000555555783874 in run_mod () at /tmp/build/80754af9/python_1553721932202/work/Python/pythonrun.c:1035
#58 0x000055555578db81 in PyRun_FileExFlags () at /tmp/build/80754af9/python_1553721932202/work/Python/pythonrun.c:988
#59 0x000055555578dd73 in PyRun_SimpleFileExFlags () at /tmp/build/80754af9/python_1553721932202/work/Python/pythonrun.c:429
#60 0x000055555578ee5f in pymain_run_file (p_cf=0x7fffffffd940, filename=0x5555558c6a40 L"/home/kinow/Development/python/workspace/cylc-doc/venv/bin/sphinx-build", 
    fp=0x55555590e210) at /tmp/build/80754af9/python_1553721932202/work/Modules/main.c:427
#61 pymain_run_filename (cf=0x7fffffffd940, pymain=0x7fffffffda50) at /tmp/build/80754af9/python_1553721932202/work/Modules/main.c:1627
#62 pymain_run_python (pymain=0x7fffffffda50) at /tmp/build/80754af9/python_1553721932202/work/Modules/main.c:2877
#63 pymain_main () at /tmp/build/80754af9/python_1553721932202/work/Modules/main.c:3038
#64 0x000055555578ef7c in _Py_UnixMain () at /tmp/build/80754af9/python_1553721932202/work/Modules/main.c:3073
#65 0x00007ffff75dfb97 in __libc_start_main (main=0x55555564aed0 <main>, argc=7, argv=0x7fffffffdba8, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, 
    stack_end=0x7fffffffdb98) at ../csu/libc-start.c:310
#66 0x0000555555734122 in _start () at ../sysdeps/x86_64/elf/start.S:103

```
</details>

```
Thread 1 "python" received signal SIGSEGV, Segmentation fault.
0x00007ffff581d962 in simplequeue_dealloc () at /usr/local/src/conda/python-3.7.6/Modules/_queuemodule.c:29
29	/usr/local/src/conda/python-3.7.6/Modules/_queuemodule.c: No such file or directory.
```

Not sure how that could happen. But at least appears to be some Python or C code that is crashing Sphinx—I think.

**Comment by kinow** [comment]
Upgraded to sphinx 2.4 just to see if that worked, no luck. Sphinx 3 breaks due to extensions before the segfault. :thinking:

**Comment by kinow** [comment]
Re-created the environment and it worked :tada: 

@oliver-sanders I believe it wasn't the sphinx dependency, but one of the other sphinx* ones that could be causing it.

**Comment by kinow** [comment]
One warning building, but not a blocker I think as it still builds fine.

```
reading sources... [100%] user-guide/writing-suites                                                                                                                            
/home/kinow/Development/python/workspace/cylc-doc/src/reference/config/global.rst:947: WARNING: Unknown directive type "todo".

.. todo::
   Is it conceivable that different remote task hosts at the same site
   might see the suite host differently? If so we would need to be able
   to override the target in suite configurations.
looking for now-outdated files... none found

```

**Pull request merged**